File: section24#Canny/C++/canny.cpp
#include <string>
#include <iostream>
#include <opencv2/opencv.hpp>

using namespace std;
using namespace cv;

static bool cannyTest(Mat imgOri, bool needResize = false);

int main()
{
	Mat imgOri = imread("Gril.jpg");
	if (imgOri.empty())
	{
		cout << "Cannot find this picture!" << endl;
		return false;
	}
	cannyTest(imgOri);

	return true;
}

static bool cannyTest(Mat imgOri, bool needResize)
{
	Mat imgEdge;

	imshow("Canny Original", imgOri);

	if (1 != imgOri.channels())
	{
		cvtColor(imgOri, imgOri, COLOR_BGR2GRAY);
	}

	if (true == needResize)
	{
		resize(imgOri, imgOri, Size(imgOri.cols/2, imgOri.rows/2), 0, 0, INTER_LINEAR);
		//convertScaleAbs(imgOri, imgOri);
	}

	GaussianBlur(imgOri, imgOri, Size(5, 5), 1);

	Canny(imgOri, imgEdge, 20, 50, 3, false);
	imshow("Canny Edge", imgEdge);

	waitKey(0);
	return true;
}

File: section24#Canny/Python/canny.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# author：lowkeyway time:8/27/2019

import sys
import cv2 as cv
import numpy as np

def cannyTest(imgOri, needResize = False, needBlur = False):
    cv.imshow("cannyTest imgOri", imgOri)
    if needResize:
        imgOri = cv.resize(imgOri, (int(imgOri.shape[1]/2), int(imgOri.shape[0]/2)))
    if needBlur:
        imgOri = cv.GaussianBlur(imgOri, (11, 11), 1)


    imgEdge = cv.Canny(imgOri, 50, 100)
    cv.imshow("cannyTest imgEdge", imgEdge)

    imgEdgeSrc = cv.bitwise_and(imgOri, imgOri, mask=imgEdge)
    cv.imshow("cannyTest imgEdgeSrc", imgEdgeSrc)

    h, w = imgOri.shape[:2]
    result = np.zeros([h, w*2, 3], dtype=imgOri.dtype)
    result[0:h, 0:w, :] = imgOri
    result[0:h, w:2*w, :] = imgEdgeSrc
    cv.putText(result, "imgOri", (10, 30), cv.FONT_ITALIC, 1.0, (0, 0, 255), 2)
    cv.putText(result, "imgUnit", (10+w, 30), cv.FONT_ITALIC, 1.0, (0, 0, 255), 2)
    cv.imshow("cannyTest Unite", result)

    cv.waitKey(0)

def main_func(argv):
    imgGirl = cv.imread("Gril.jpg")
    cannyTest(imgGirl, needResize=True, needBlur=False)


if __name__ == '__main__':
    main_func(sys.argv)